### DISCUSSION TITLE
x-minio-force-delete or 'mc rb' when delete non-empty bucket

### DISCUSSION THREAD

**Discussion by Clara12062** [minio/minio#20367] [Q&A]
Hi~
Currently, there is a bucket with around 10,000 objects in it. Now I want to quickly delete the data in the bucket. I also found that Minio provides a request parameter for forcefully deleting a bucket: `x-minio-force-delete`.
When using the minio-go interface to delete, I found that when there are many objects, it will slow down and become sluggish.
And then Using `mc rb --force ALIAS/BUCKET` to delete takes a long time. I found that even though the `--force` parameter was added, it first deletes the objects and then calls the forceful deletion of the bucket again. And this way is just slow and won't get stuck.

And my question is: Is it not recommended to use `x-minio-force-delete` to delete a bucket when there is a lot of data in it?

**Comment by marktheunissen** [comment]
Hi @Amber-976 `mc rb` will use the header `x-minio-force-delete` too, so whether you call `mc` or use the `minio-go` SDK, the result is the same. 

> it will slow down and become sluggish.

Please be much more specific... what slows down? Can you measure "becomes sluggish"? What are the logs, steps to reproduce?

**Reply by Clara12062** [reply]
Thank you for your reply!

I did the following, under a bucket, with 300,000 + objects uploaded (although this is not the recommended way to use it 😂  ).
Then using `time mc rb ALIAS/BUCKET`, it takes about 7min.
```sh
[root@xos-1x6e1kd5 ~]# time  mc rb --force minio/dxtest
Removed `minio/dxtest` successfully.

real	7m42.580s
user	0m19.359s
sys	0m1.039s

```
Then again there are 30,000 + objects in bucket to call the interface:

```golang
opts := minio.RemoveBucketOptions{
ForceDelete: true,
}

err = minioClient.RemoveBucketWithOptions(context.Background(), "dxtest", opts)
if err ! = nil {
log.Fatalln(err)
}
```
The results are different.
```sh
[root@xos-1x6e1kd5 ~]# time /root/dx/delminio 
2024/09/03 16:50:18 Bucket dxtest deleted successfully

real	0m42.416s
user	0m0.004s
sys	0m0.003s
```

Using `x-minio-force-delete` directly is faster. The previous problem could have been a virtual disk issue.

That is, in production, both are fine, right?

**Reply by marktheunissen** [reply]
This could be for various reasons - when MinIO deletes a bucket like that, it first moves the files into a kind-of "trash", then deletes them asynchronously. So any testing has to be done carefully to make sure you are measuring the right thing. 

7 minutes is very long though, I suppose with 300k objects that's how long it takes. What is your hardware? 

Anyway both options are fine to use in production, although you probably aren't deleting such large buckets on regular basis?

**Reply by Clara12062** [reply]
Thanks for your reply again！

> 7 minutes is very long though, I suppose with 300k objects that's how long it takes. What is your hardware?

SSD。Maybe I have too few disks? The minio cluster is 1server4disks.
And `mc rb` seems to batch delete all objects in the bucket first, and then use `DeleteForce: true` last.
`x-minio-force-delete` directly renames the entire directory to trash, and then removeall. 
Is it possible that mc is executing slowly, not minio server side?

> Anyway both options are fine to use in production, although you probably aren't deleting such large buckets on regular basis?

yes. This is just for testing. I'd like to find out what the difference is between the two.

**Reply by harshavardhana** [reply]
you can check your average DeleteMultipleObjects response time via `mc admin trace`

**Reply by harshavardhana** [reply]
1000 objects per bucket delete takes 1s then 300k will take 300 * 1s
